### PR TITLE
DM-35102

### PR DESCRIPTION
**Pull request by pkubanek**
Improve (Air Compressor) GUIs:
- DataFormButton
- SummaryStateLabel
- Pressure label
- Improved displays
- User can select & plot values

### CODE REVIEW

File: python/lsst/ts/cRIOpy/GUI/CompressorsPageWidget.py [hunk 1]
```diff
@@ -177,139 +192,146 @@ def _addButton(text):
...
-            "Start after power-loss",
-            OnOffLabel(self.compressor.status, "startAfterPowerLoss"),
+        statusLayout = QVBoxLayout()
+        statusLayout.addWidget(
+            DataFormWidget(
+                self.compressor.status,
+                [
+                    ("Ready to start", OnOffLabel(field="readyToStart")),
+                    ("Operating", PowerOnOffLabel(field="operating")),
+                    ("Start Inhibit", OnOffLabel(field="startInhibit")),
```
Comment (teweitsai) [inline python/lsst/ts/cRIOpy/GUI/CompressorsPageWidget.py:275]: Based on other part, maybe "Start inhibit"?

File: python/lsst/ts/cRIOpy/GUI/CustomLabels.py [hunk 2]
```diff
@@ -42,8 +44,15 @@
     "Moment",
     "Mm",
     "Arcsec",
+    "Ampere",
     "Percent",
     "Volt",
+    "RPM",
+    "PressureBar",
+    "PressuremBar",
```
Comment (teweitsai) [inline python/lsst/ts/cRIOpy/GUI/CustomLabels.py:52]: "PressuremBar"??? Or you want to say "pressure in mBar"?

File: python/lsst/ts/cRIOpy/GUI/CustomLabels.py [hunk 3]
```diff
@@ -123,8 +134,8 @@ class UnitLabel(QLabel):
     fmt : `str`, optional
         Format string. See Python formatting function for details. Defaults to
         'd' for decimal number.
-    unit : `astropy.units`, optional
-        Variable unit. Default is None - no unit
+    unit : `astropy.units|str`, optional
```
Comment (teweitsai) [inline python/lsst/ts/cRIOpy/GUI/CustomLabels.py:137]: `astropy.units` or `str`

File: python/lsst/ts/cRIOpy/GUI/CustomLabels.py [hunk 4]
```diff
@@ -322,35 +344,127 @@ def __init__(self, fmt="0.02f", is_warn_func=None, is_err_func=None):
...
+
+class PressureBar(DataLabel):
+    """Display pressure in bar and psi"""
+
+    def __init__(self, signal=None, field=None):
+        super().__init__(signal, field)
+        self.unit_name = "bar"
+
+    def setValue(self, value):
+        psi = value * 14.5038
```
Comment (teweitsai) [inline python/lsst/ts/cRIOpy/GUI/CustomLabels.py:435]: Maybe give a variable name to 14.5038 for the readability?

File: python/lsst/ts/cRIOpy/GUI/CustomLabels.py [hunk 5]
```diff
@@ -322,35 +344,127 @@ def __init__(self, fmt="0.02f", is_warn_func=None, is_err_func=None):
...
+
+
+class PressuremBar(DataLabel):
+    def __init__(self, signal=None, field=None):
+        super().__init__(signal, field)
+        self.unit_name = "mbar"  # this is only for display
+
+    def setValue(self, value):
+        value /= 1000.0
+        psi = value * 14.5038
```
Comment (teweitsai) [inline python/lsst/ts/cRIOpy/GUI/CustomLabels.py:446]: Maybe give the variable names to 1000.0 and 14.5038 for the readability?

File: python/lsst/ts/cRIOpy/GUI/DataFormWidget.py [hunk 6]
```diff
@@ -58,3 +61,54 @@ def _process_signal(self, data):
...
+    def mousePressEvent(self, ev):
+        if self._timeChart is not None:
+            child = self.childAt(ev.pos())
+            if child is not None:
+                self._timeChart.topicSelected.emit(child)
+
+
+class DataFormButton(QPushButton):
+    """
+    Creates button displaying overall status. On click, . Update fields on signal with new values.
```
Comment (teweitsai) [inline python/lsst/ts/cRIOpy/GUI/DataFormWidget.py:74]: "On click, ." ???

File: python/lsst/ts/cRIOpy/GUI/TimeChart.py [hunk 7]
```diff
@@ -105,18 +111,26 @@ def __init__(self, items, maxItems=50 * 30, updateInterval=0.1):
...
-                    self._addSerie(d, axis[0])
-            self._caches.append(TimeCache(maxItems, data))
+        self._createCaches(items, maxItems)
+
+        self._attachSeries()
+
+    def _addSerie(self, name, axis):
+        s = QtCharts.QLineSeries()
+        s.setName(name)
+        # s.setUseOpenGL(True)
```
Comment (teweitsai) [inline python/lsst/ts/cRIOpy/GUI/TimeChart.py:121]: Do you still need this commented code?

File: python/lsst/ts/cRIOpy/GUI/TimeChart.py [hunk 8]
```diff
@@ -221,6 +248,46 @@ def clearData(self):
         super().removeAllSeries()
 
 
+class UserSelectedTimeChart(TimeChart):
+    """
+    Signals
+    -------
```
Comment (teweitsai) [inline python/lsst/ts/cRIOpy/GUI/TimeChart.py:254]: Not sure I understand this:

```
Signals
-------
```

Do you try to say "Attributes"?

File: python/lsst/ts/cRIOpy/GUI/TimeChart.py [hunk 9]
```diff
@@ -221,6 +248,46 @@ def clearData(self):
...
+            for n in t.DataType().get_vars().keys():
+                if n == name:
+                    if self._signal is not None:
+                        self._signal.disconnect(self._appendData)
+
+                    self._signal = s
+                    self._name = name
+
+                    self._signal.connect(self._appendData)
+                    break
```
Comment (teweitsai) [inline python/lsst/ts/cRIOpy/GUI/TimeChart.py:284]: Is there any way to reduce the indentation here?

**Review by teweitsai — APPROVED** [review]
Looks good. Thanks for the improvement!

File: python/lsst/ts/cRIOpy/GUI/CompressorsPageWidget.py [hunk 10]
```diff
@@ -177,139 +192,146 @@ def _addButton(text):
...
+                    ),
+                    (
+                        "Stage 1 output pressure",
+                        PressuremBar(field="stage1OutputPressure"),
+                    ),
+                    ("Line pressure", PressuremBar(field="linePressure")),
+                    (
+                        "Stage 1 output temperature",
+                        DataDegC(field="stage1OutputTemperature", fmt=".0f"),
+                    ),
```
[the remark this answers is not in the document]
Comment (pkubanek) [reply python/lsst/ts/cRIOpy/GUI/CompressorsPageWidget.py:246]: I looked into modbus map (https://confluence.lsstcorp.org/display/LTS/Datasheets?preview=/154697856/182423947/Modbus%20Table%20-%20CompAir%20-%20Delcos%20XL%20-%20L%26DH-Range%2014.12.2020.pdf), but there isn't anything about stage n (n>1).

File: python/lsst/ts/cRIOpy/GUI/CustomLabels.py (hunk 5 above)
Comment (pkubanek) [reply python/lsst/ts/cRIOpy/GUI/CustomLabels.py:446]:
```python
        mbar = value * u.mbar
        bar = (mbar).to(u.bar).value
        psi = mbar.to(u.imperial.psi).value
```
is even nicer.

File: python/lsst/ts/cRIOpy/GUI/TimeChart.py (hunk 7 above)
Comment (pkubanek) [reply python/lsst/ts/cRIOpy/GUI/TimeChart.py:121]: It wasn't working at some stage. Now it seems to be happy and looks like graphing is much faster - uncommented it.

**Pull request merged**